srtcore/buffer_tools.cpp

```diff
@@ -273,5 +273,66 @@ int CSndRateEstimator::incSampleIdx(int val, int inc) const
     return val;
 }
 
+CMovingRateEstimator::CMovingRateEstimator()
+    : m_tsFirstSampleTime(sync::steady_clock::now())
+    , m_iFirstSampleIdx(0)
+    , m_iCurSampleIdx(0)
+    , m_iRateBps(0)
+    , m_Samples(NUM_PERIODS)
+{
+    resetRate(0, NUM_PERIODS);
+}
+
+void CMovingRateEstimator::addSample(int pkts, double bytes)
+{
+    const time_point now             = steady_clock::now();
+    const int        iSampleDeltaIdx = int(count_milliseconds(now - m_tsLastSlotTimestamp) / SAMPLE_DURATION_MS);
+
+    if (iSampleDeltaIdx == 0)
+    {
+        m_Samples[m_iCurSampleIdx].m_iBytesCount += bytes;
+        m_Samples[m_iCurSampleIdx].m_iPktsCount += pkts;
+    }
+    else
+    {
+        if ((m_iCurSampleIdx + iSampleDeltaIdx) < NUM_PERIODS)
+            resetRate(m_iCurSampleIdx + 1, m_iCurSampleIdx + iSampleDeltaIdx);
+        else
+        {
+            int loopbackDiff = m_iCurSampleIdx + iSampleDeltaIdx - NUM_PERIODS;
+            resetRate(m_iCurSampleIdx + 1, NUM_PERIODS);
+            resetRate(0, loopbackDiff);
+        }
+
+        m_iCurSampleIdx                          = ((m_iCurSampleIdx + iSampleDeltaIdx) % NUM_PERIODS);
+        m_Samples[m_iCurSampleIdx].m_iBytesCount = bytes;
+        m_Samples[m_iCurSampleIdx].m_iPktsCount  = pkts;
+
+        m_tsLastSlotTimestamp += milliseconds_from(SAMPLE_DURATION_MS * iSampleDeltaIdx);
+
+        computeAverageValue();
+    }
 }
 
+void CMovingRateEstimator::resetRate(int from, int to)
+{
+    for (int i = max(0, from); i < min(int(NUM_PERIODS), to); i++)
+        m_Samples[i].reset();
+}
+
+void CMovingRateEstimator::computeAverageValue()
+{
+    const time_point now           = steady_clock::now();
+    const int        startDelta    = count_milliseconds(now - m_tsFirstSampleTime);
+    const bool       isFirstPeriod = startDelta < (SAMPLE_DURATION_MS * NUM_PERIODS);
+    int              newRateBps    = 0;
+
+    for (int i = 0; i < NUM_PERIODS; i++)
+        newRateBps += (m_Samples[i].m_iBytesCount + (CPacket::HDR_SIZE * m_Samples[i].m_iPktsCount));
+
+    if (isFirstPeriod)
+        newRateBps = newRateBps * 1000 / startDelta;
```

Comment:
!!! Division by zero not prevented!

Reply:
It is :) We don't go in the computeAverageValue method if iSampleDeltaIdx == 0.

Reply:
```shell
[ RUN      ] TestSocketOptions.LossMaxTTL
/home/travis/.travis/functions: line 109:  7133 Floating point exception(core dumped) ./test-srt -disable-ipv6
[New LWP 7133]
[New LWP 7746]
[New LWP 7749]
[New LWP 7750]
[New LWP 7745]
[New LWP 7747]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `./test-srt -disable-ipv6'.
Program terminated with signal SIGFPE, Arithmetic exception.
#0  0x000000000087cafc in srt::CMovingRateEstimator::computeAverageValue (this=0x7fe6bc005cb8) at /home/travis/build/Haivision/srt/srtcore/buffer_tools.cpp:334
334	        newRateBps = newRateBps * 1000 / startDelta;
[Current thread is 1 (Thread 0x7fe6c9ee9740 (LWP 7133))]
#0  0x000000000087cafc in srt::CMovingRateEstimator::computeAverageValue (this=0x7fe6bc005cb8) at /home/travis/build/Haivision/srt/srtcore/buffer_tools.cpp:334
#1  0x000000000087c94d in srt::CMovingRateEstimator::addSample (this=0x7fe6bc005cb8, pkts=0, bytes=0) at /home/travis/build/Haivision/srt/srtcore/buffer_tools.cpp:313
#2  0x00000000008d80bb in srt::stats::Sender::updateRate (this=0x7fe6bc005c18, pkts=0, bytes=0) at /home/travis/build/Haivision/srt/srtcore/stats.h:174
#3  0x00000000008b934a in srt::CUDT::bstats (this=0x7fe6bc000938, perf=0x7ffd64377750, clear=false, instantaneous=false) at /home/travis/build/Haivision/srt/srtcore/core.cpp:7577
#4  0x000000000085bfe3 in srt::CUDT::bstats (u=412356694, perf=0x7ffd64377750, clear=false, instantaneous=false) at /home/travis/build/Haivision/srt/srtcore/api.cpp:4338
```

Reply:
> It is :) We don't go in the computeAverageValue method if iSampleDeltaIdx == 0.

The division by 0 is made using the value that is calculated out of the present time taken exactly in this function, so there's even no possibility that you could control this value this way.